### DISCUSSION TITLE
Support Redis for repository caching

### DISCUSSION THREAD

**Discussion by colinodell** [renovatebot/renovate#16497] [Suggest an Idea]
## Challenge

The [repository cache](https://github.com/renovatebot/renovate/pull/6589) is currently stored within the global `cacheDir`.  If a user has multiple instances of Renovate workers running in a k8s cluster, it seems this repository cache can only be shared between them using shared k8s volume.  While this probably works, mounting a volume on a _subdirectory_ of `/tmp/renovate` to share JSON-structured cache data feels a bit clunky, especially when Renovate already some functionality to share JSON data via Redis.  (I don't want to mount the entire `cacheDir` for performance reasons)

## Proposal

What if the repository cache supported Redis as an optional alternative to the file system?  While I understand [the rationale behind excluding the repository cache from the global/package cache](https://github.com/renovatebot/renovate/pull/6589#issuecomment-650744865), what if users could configure a separate Redis URL for this particular cache?  From what I can tell, the structure of `RepoCacheBase` would easily support this, but `redis.ts` would need some tweaks to support multiple instances.

For configuration, perhaps an optional new `repositoryCacheRedisUrl` global config parameter could be added to enable this?  If set, that Redis instance would be used instead of the local cache.

Users would then have the flexibility to cache to a completely separate Redis instance, or a separate database on their current instance.  (Or they could completely ignore best practices and use the global cache, ignoring warnings we'd probably put in the docs for this new option at their own peril.)

WDYT?  Does something like this seem reasonable?  If not, do you have any advice on whether a shared volume or something else might be a better fit to ensure this cached data is shared?

**Comment by rarkins** [comment]
We could potentially refactor `repositoryCache` to be a string where possible values include:
- `local` (or maybe `file`)
- `redis://..........` (your case)
- `s3://........` (future)

We should make sure that we use a namespace for the redis keys which is already compatible with the package cache, in case someone wants to reuse the same Redis instance.

One thing to beware of is that in future we may grow the data we put into repository cache so you'd be responsible for making sure you don't run out of memory/storage.

**Reply by viceice** [reply]
i like the S3 idea.

**Reply by rarkins** [reply]
@colinodell was this a feature you were interested in implementing yourself?

**Reply by colinodell** [reply]
Yeah, I'd be happy to take a stab at this! I'll aim to have a PR ready for review in next week or two :)

**Reply by setchy** [reply]
some solid foundations now in place on which to add other cache client support - https://github.com/renovatebot/renovate/pull/17146

**Reply by colinodell** [reply]
Thanks for the heads-up! I'll rebase my implementation (#16735) onto #17146 later this week (or weekend).

**Reply by setchy** [reply]
@colinodell - anything I could help with re: the rebase?

**Comment by mieliespoor** [comment]
Just to make sure I understand the caching here correctly... Does this mean that you could have two renovate processes running and because of the shared cache, the second won't iterate over the repositories that the first did?

**Reply by mieliespoor** [reply]
With 10 repositories, you then see this:
proc1 does repo's:
 - 1,2,4,7,8
proc2 does repo's:
 - 3,5,6,9,10

This is very much over simplified, but hopefully illustrates the point.

**Reply by rarkins** [reply]
This feature alone won't make any changes to *which* repos each renovate instance processes. It simply would enable a shared cache
